### DISCUSSION TITLE
unauthenticated proxy for iot device sdk

### DISCUSSION THREAD

**Discussion by xtx2017** [Azure/azure-iot-sdk-java#1443] [Q&A] [answered]
Hi, I want set an unauthenticated proxy for IoT device SDK.
I find DeviceClient class has setProxySettings method.
ProxySettings class has a constructor for unauthenticated proxy.
```java
public ProxySettings(Proxy proxy){
    this(proxy, null, null);
}
```
However, the inner called constructor throw Exception.
```java
public ProxySettings(Proxy proxy, final String username, final char[] password) {
    ...
    if (username != null && username.isEmpty(); {
        throw new IllegalArgumentException("Username cannot be an empty string");
    }
    ...
}
```
How can I set unauthenticated proxy for IoT device SDK?

p.s. 
I wonder that the proxy setting for IoT service SDK is only for unauthenticated proxy different from device SDK.
(com.microsoft.azure.sdk.iot.service.devicetwin.ProxyOptions does not have username and password setting)

**Comment by timtay-microsoft** [comment, marked as answer]
The constructor 

```java
public ProxySettings(Proxy proxy){
    this(proxy, null, null);
}
```

allows you to create your client with an unauthenticated proxy. The IllegalArgumentException that you have highlighted here

```java
    if (username != null && username.isEmpty(); {
        throw new IllegalArgumentException("Username cannot be an empty string");
    }
```

assumes that if the username is null, then no authentication will be used when connecting to the proxy. It just doesn't allow for the username to be an empty string since we thought that was an ambiguous use case.

> I wonder that the proxy setting for IoT service SDK is only for unauthenticated proxy different from device SDK.
(com.microsoft.azure.sdk.iot.service.devicetwin.ProxyOptions does not have username and password setting)

Currently the service client only supports unauthenticated proxies, yes.

Hope that helps!
-Tim

**Comment by xtx2017** [comment]
Hi, Tim.
Thank you for your reply. I have a simple mistake in my code.
I misunderstood the exception occurs on calling  ProxySettings(Proxy proxy) and the exception condition is "username == null || username.isEmpty()". LOL
